rclcpp/test/rclcpp/test_loaned_message.cpp

```diff
@@ -79,5 +79,7 @@ TEST_F(TestLoanedMessage, release) {
 
   ASSERT_EQ(42.0f, msg->float64_value);
 
+  msg->~MessageT();
+  pub->get_allocator()->deallocate(msg, 1);
```

Comment:
This seems like a weird API.  That is, is this the API a user of the borrow API is supposed to use?  If not, what is a user who is borrowing a message, but then calling `release` on it, supposed to call to free the memory?

Reply:
The release of LoanedMessage is to `Release ownership of the ROS message instance`.
https://github.com/ros2/rclcpp/blob/0313417f0213f1741465978c91ff18ba4f99ab55/rclcpp/include/rclcpp/loaned_message.hpp#L192
After `release`, the destructor is not going to destroy the message.
https://github.com/ros2/rclcpp/blob/0313417f0213f1741465978c91ff18ba4f99ab55/rclcpp/include/rclcpp/loaned_message.hpp#L135-L137

So user need to do the same way as
https://github.com/ros2/rclcpp/blob/0313417f0213f1741465978c91ff18ba4f99ab55/rclcpp/include/rclcpp/loaned_message.hpp#L149-L151

Reply:
OK, I will not use this weird API to destroy it.
Update it with
```
ASSERT_EQ(RCL_RET_OK, rcl_return_loaned_message_from_publisher(pub.get_publisher_handle().get(), msg));
```

Reply:
That seems like a reasonable API to add.  However, I'll ping @mjcarroll and @wjwwood who were initially involved with this feature to see what the expected sequence is supposed to be.

Reply:
Grad to hear that you also think it's a reasonable API. I'll keep my eyes on it.

Reply:
thanks for retagging me here. I did not really see this within all GH notifications.

I think I've added that previously to mimic the behavior of a unique_ptr. Looking roughly through the code base, I think it's safest to simply deprecate this function or directly completely remove it. I don't really see a good use-case for it and as shown here might lead to more confusion than benefits.

Reply:
@wjwwood @Karsten1987

Thanks for your response. I'll remove this method and the test case.

Reply:
@clalancette 
If all of you agree to add a unique_ptr version of release pointed by @wjwwood, I'll add it later.

Reply:
> I don't really see a good use-case for it and as shown here might lead to more confusion than benefits.

I tend to agree with this atm. If we really want to keep it, then I think we should capture the destruction in the `unique_ptr` so it is done safely. Otherwise, I think we should just get rid of it.

Reply:
After reconsideration, I think this API is useful for some advanced users.